cs/Serializer/VowpalWabbitDefaultMarshaller.cs

```diff
@@ -0,0 +1,358 @@
+            {
+                return;
+            }
+
+            foreach (DictionaryEntry item in value)
+            {
+                context.NamespaceBuilder.AddFeature(
+                    context.VW.HashFeature(Convert.ToString(item.Key), ns.NamespaceHash),
+                    (float)Convert.ToDouble(item.Value, CultureInfo.InvariantCulture));
+            }
+
+            if (context.StringExample == null)
+            {
+                return;
+            }
+
+            foreach (DictionaryEntry item in value)
+            {
+                context.AppendStringExample(
+                    feature.Dictify,
+                    " {0}:" + (context.VW.Settings.EnableStringFloatCompact ? "{1}" : "{1:E20}"),
+                    Convert.ToString(item.Key),
+                    (float)Convert.ToDouble(item.Value, CultureInfo.InvariantCulture));
+            }
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="context">The marshalling context.</param>
+        /// <param name="ns">The namespace description.</param>
+        /// <param name="feature">The feature description.</param>
+        /// <param name="value">The actual feature value.</param>
+        public void MarshalFeature(VowpalWabbitMarshalContext context, Namespace ns, Feature feature, IEnumerable<string> value)
+        {
+            Contract.Requires(context != null);
+            Contract.Requires(ns != null);
+            Contract.Requires(feature != null);
+
+            if (value == null)
+            {
+                return;
+            }
+
+            foreach (var item in value)
+            {
+                context.NamespaceBuilder.AddFeature(context.VW.HashFeature(item.Replace(' ', '_'), ns.NamespaceHash), 1f);
+            }
+
+            if (context.StringExample == null)
+            {
+                return;
+            }
+
+            foreach (var item in value)
+            {
+                context.AppendStringExample(feature.Dictify, " {0}", item);
+            }
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="context">The marshalling context.</param>
+        /// <param name="ns">The namespace description.</param>
+        /// <param name="featureVisits"></param>
+        public void MarshalNamespace(VowpalWabbitMarshalContext context, Namespace ns, Action featureVisits)
+        {
+            try
+            {
+                // the namespace is only added on dispose, to be able to check if at least a single feature has been added
+                context.NamespaceBuilder = context.ExampleBuilder.AddNamespace(ns.FeatureGroup);
+
+                var position = 0;
+                var stringExample = context.StringExample;
+                if (context.StringExample != null)
+                {
+                    position = stringExample.Append(ns.NamespaceString).Length;
+                }
+
+                featureVisits();
```

Comment:
It visit all the features, but if none of the visits generated a feature (in native terms), the namespace must not be added.